pages/api/auth/register.ts

```diff
@@ -0,0 +1,100 @@
+import axios from 'axios';
+import { NextApiRequest, NextApiResponse } from 'next';
+import { handleError } from '~/backend/handleError';
+import { HttpError } from '~/backend/HttpError';
+import { CUSTOM_HEADER } from '~/middleware';
+import type { CreateUser, User } from '~/shared/types';
+import { emailRegex, passwordRegex } from '~/utils/regexPatterns';
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse,
+) {
+  try {
+    if (req.method !== 'POST') {
+      throw new HttpError('Invalid request method!', 405);
+    }
+    const requestBody = req.body as CreateUser;
+    if (
+      !requestBody.email ||
+      !requestBody.firstName ||
+      !requestBody.lastName ||
+      !requestBody.password
+    ) {
+      throw new HttpError('Invalid request body!', 400);
+    }
+    isFirstNameValid(requestBody.firstName);
+    isLastNameValid(requestBody.lastName);
+    isEmailValid(requestBody.email);
+    isPasswordValid(requestBody.password);
+
+    if (CUSTOM_HEADER.key === undefined || CUSTOM_HEADER.value === undefined) {
+      throw new HttpError('Server error!', 500);
+    }
+    const userCreationRequest = (await (
+      await axios.post(
+        process.env.NODE_ENV === 'production'
+          ? '/api/users'
+          : 'http://localhost:3000/api/users',
+        {
+          email: requestBody.email,
+          firstName: requestBody.firstName,
+          lastName: requestBody.lastName,
+          password: requestBody.password,
+        } as CreateUser,
+        {
+          headers: {
+            [CUSTOM_HEADER.key]: CUSTOM_HEADER.value,
+          },
+        },
+      )
+    ).data) as User;
```

Comment:
Kun käyttäjä rekistöröityy esim. `http://localhost:3000/register`-sivulla, se lähettää tähän tiedostoon ensimmäisenä POST-requestin, requesti käydään läpi tässä tiedostossa ja katsotaan onko kaikki tiedot kunnossa. Jos tiedot ovat kunnossa, backend lähettää `/api/users` API:in POST-requestin samalla tavalla, mutta varmistetuilla tiedoilla.

Tässä kohtaa tulee ongelma, koska `middleware` haluaa `Origin`-headerin (oletettavasti CSRF-hyökkäystä vastaan). `Origin`-headeria backend ei saa, jolloin `middleware` heittää `403 Forbidden` -errorin, koska backendiä ei ole "tunnistettu"

Tähän kohtaan tein "kiertoreitin", joka ei ole tietoturvallisin ratkaisu, mutta backend yrittää lähettää `Users`-API:lle oman muokatun `header`:in, jonka `middleware` tarkistaa. Jos oma `header` täsmää ennalta annettuihin arvoihin, antaa `middleware` pääsyn eteenpäin tuosta ensimmäisestä tarkistuksesta.

Olisiko siis parhain, että käyttäjä luotaisiikin `/pages/api/auth/register.ts`-tiedostossa, eikä lähetettäisi pyyntö API:lle?

Tämä on melko vaikea selittää mitä haen takaa, mutta yritin parhaani :sweat_smile: Laitan Discordissa vielä esimerkin noista "salaisista" arvoista

Reply:
Omaa backendiä ei kuulu kutsua POST-requestilla, vaan suoraan funtiota. Eli eriytä `/api/users` endpointin se koodi minkä haluat ajaa täältä käsin omaan funktioon, ja importtaa ja kutsu suoraan sitä funktiota.